Improve base_score assignment

lsdxp opened the pull request:
Refactor base_score assignment to handle string and collection types. in xgboost 310, the base_score return an array instead of a simple float

## Overview

Description of the changes proposed in this pull request:
BUG is reported in the issue #4184  due to changes in xgboost base_score return data type.

Diff of the file in xgboost is here: [Xgboost base_score](https://github.com/dmlc/xgboost/pull/11651/files#diff-92691cadfd8852da44fa622dceb9ada7c352a0b8a62e158ee2e87a95feeab6baR85-R87
)
from:

 Line 86 float base_score{ObjFunction::DefaultBaseScore()};

to

Line 86  common::ParamArray<float, true> base_score{"base_score", ObjFunction::DefaultBaseScore()};



## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.

lsdxp committed:
Improve base_score assignment 

Refactor base_score assignment to handle string and collection types. in xgboost 310, the base_score return an array instead of a simple float

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

CloseChoice commented on shap/explainers/_tree.py:
```diff
@@ -2101,13 +2102,21 @@ def __init__(self, xgb_model) -> None:
+        if isinstance(base_score, str):
+            try:
+                base_score = ast.literal_eval(base_score)
+            except Exception:
```
Could you please catch the correct exception (I believe it's a value error) and then throw an exception that is infromative for the user?

Something like: "expected the base_score to contain a list or float, received {base_score}"?

lsdxp replied:
Added the exception message and restricted the base_score to be certain types to match the message

CloseChoice reviewed (changes requested):
Thanks for the PR, this looks good so far. 

Not quite sure if we'll need a test, this depends on this pipeline on the master branch: https://github.com/shap/shap/actions/runs/18836154340

If this passes successfully, we'll need one, otherwise this PR fixes the current pipeline

lsdxp committed:
Improve base_score validation in _tree.py

Add error handling for invalid base_score types and throw out clear exception message

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lsdxp commented:
> Thanks for the PR, this looks good so far.
> 
> Not quite sure if we'll need a test, this depends on this pipeline on the master branch: https://github.com/shap/shap/actions/runs/18836154340
> 
> If this passes successfully, we'll need one, otherwise this PR fixes the current pipeline

looks tests fail except with python3.9, assume current xgboost related tests already covering this part?

CloseChoice commented:
> > Thanks for the PR, this looks good so far.
> > Not quite sure if we'll need a test, this depends on this pipeline on the master branch: https://github.com/shap/shap/actions/runs/18836154340
> > If this passes successfully, we'll need one, otherwise this PR fixes the current pipeline
> 
> looks tests fail except with python3.9, assume current xgboost related tests already covering this part?
the latest xgboost release probably does not support python 3.9, therefore they pass. I let the CI run, let's see if this fixes the currently failing pipelines

CloseChoice reviewed (approved):
LGTM, thanks a lot for the quick fix, great work!
